zcm/julia/ZCM.jl

```diff
@@ -0,0 +1,334 @@
+module ZCM
+
+# RRR / Note: Julia requires that the memory layout of the C structs is consistent
+#             between their definitions in zcm headers and this file
+
+# C ptr types for types that we don't need the internals of
+module Native
+
+type Zcm
+end
+
+type Sub
+end
+
+type UvSub
+end
+
+type RecvBuf
+    recv_utime ::Int64;
+    zcm        ::Ptr{Native.Zcm};
+    data       ::Ptr{UInt8};
+    data_size  ::UInt32;
+end
+
+type EventLog
+end
+
+type EventLogEvent
+    eventnum  ::Int64;
+    timestamp ::Int64;
+    chanlen   ::Int32;
+    datalen   ::Int32;
+    channel   ::Ptr{UInt8};
+    data      ::Ptr{UInt8};
+end
+
+end
+
+# RRR: make variable naming consistent
+# RRR: improve orgainzation and make header comments more visible
+# Exported Objects and Methods
+type RecvBuf
+    rbuf  ::Ptr{Native.RecvBuf};
+    utime ::Int64;
+    data  ::Array{UInt8};
+
+    function RecvBuf(rbuf::Ptr{Native.RecvBuf})
+        instance  = new();
+        instance.rbuf = rbuf;
+
+        if (rbuf != C_NULL)
+            loadedRbuf = unsafe_load(rbuf);
+
+            instance.utime = loadedRbuf.recv_utime;
+
+            if (loadedRbuf.data != C_NULL)
+                instance.data = unsafe_wrap(Array, loadedRbuf.data, loadedRbuf.data_size);
+            else
+                instance.data = [];
+            end
+        end
+
+        # destruction of the underlying Native.RecvBuf pointer is handled by the C lib internally
+
+        return instance;
+    end
+end
+export RecvBuf;
+
+type __zcm_handler
+    handler::Any;
+    usr::Any;
+
+    # handler must have the function or functor prototype :
+    #   (rbuf::RecvBuf, channel::String, msg::MsgType, usr)
+    __zcm_handler(handler, usr) = new(handler, usr);
```

Comment:
This constructor is unnecessary. In addition, I'm not sure that the subscription interface is being handled in the best way. Particularly, passing the `usr` field around as a pointer and then doing `unsafe_pointer_to_objref` is scary and unnecessary. Instead, I would strongly suggest just asking users to pass a closure over whatever user variables they're interested in. That's what I do in https://github.com/rdeits/LCMCore.jl 

So, rather than a user doing (pseudocode):

```julia
subscribe(zcm, "FOO", foo_t, callback, usr)
```

they would just construct a callback that closes over their user-data and subscribe that:

```
callback = (buf, channel) -> myfunction(buf, channel, usr)
subscribe(zcm, "FOO", foo_t, callback)
```

That way you're letting Julia do the work of managing the user data and ensuring it doesn't get garbage collected. 

In fact, I've been pretty happy with the subscription system in LCMCore.jl, and I think it could probably be ported entirely into ZCM. Want me to give it a try?

Reply:
Also, I think that the Zcm object should actually contain the list of subscriptions. The reason is that if the `Sub` returned by `subscribe()` goes out of scope, the ZCM subscription still exists. That can result in the handler being called after the Julia `Sub` has been freed, which will crash. That's also something I can port over from LCMCore.jl